sale_loyalty_order_line_link/models/sale_order.py

```diff
@@ -175,11 +149,15 @@ def _filter_related_program_lines(self, program):
         """Discard those lines not in the program domain. With other modules changing
         the product criteria rules, we can extend this method to return the proper
         records."""
-        # No domain means that any product meets the promotion rules
-        if not program.rule_products_domain:
-            return self
-        domain = safe_eval(program.rule_products_domain)
-        products = self.mapped("product_id").filtered_domain(domain)
-        return self.filtered(
-            lambda x: x.product_id in products and not x.is_reward_line
-        )
+        programs_line = self.env["sale.order.line"]
+        for rule in program.rule_ids:
```

Comment:
Don't iterate, but append the different rules domain and make the filter at once